llvm/lib/CodeGen/GlobalISel/CSEMIRBuilder.cpp

```diff
@@ -174,6 +174,20 @@ MachineInstrBuilder CSEMIRBuilder::buildInstr(unsigned Opc,
   switch (Opc) {
   default:
     break;
+  case TargetOpcode::G_ICMP: {
+    assert(SrcOps.size() == 3 && "Invalid sources");
+    assert(DstOps.size() == 1 && "Invalid dsts");
+    LLT SrcTy = SrcOps[1].getLLTTy(*getMRI());
+
+    if (std::optional<SmallVector<APInt>> Cst =
```

Comment:
I don't see the point of using APInt in this interface. You're discarding the used result type here, and then forcing it into a 1-bit APInt. Either preserve the original result type, or just return a bool?

Reply:
There is a type check in `MachineIRBuilder::buildConstant` to make sure their bits are same. The use of `APInt` is to unify the `ConstantFoldICmp` interface to make it support both vector and scalar value.

Reply:
In that case it would accept `ArrayRef<int64_t>` I suppose, which still needs to have an explicit construction of that array because we can't have implicit cast from, say `ArrayRef<bool>` to `ArrayRef<int64_t>`. That doesn't look too much different from `ArrayRef<APInt>`.